<a href="https://colab.research.google.com/github/PRihana/Machine-learning/blob/master/24_03_2020(ML)pandas_contd.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import numpy as np
import pandas as pd

In [0]:
# numpy version check
print(np.__version__)
# Pandas version
print(pd.__version__)

1.18.2
0.25.3


In [0]:
# Read the data
data = pd.read_csv("/FL_insurance_sample.csv.txt")

In [0]:
# Checking output type
print(type(data))

<class 'pandas.core.frame.DataFrame'>


In [0]:
# To check the shape 
print(data.shape)
# to get the number of rows or instances or records or examples
print(data.shape[0])
# to get the number of columns or features or attributes
print(data.shape[1])

(9462, 18)
9462
18


In [0]:
# To get the column names
print(data.columns)
# To get the row names
print(data.index)

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')
RangeIndex(start=0, stop=9462, step=1)


In [0]:
# Datatype of each column
print(data.dtypes)
print("-----"*15)
# what is output type for data.dtypes
print(type(data.dtypes))
print("-----"*15)
# Access the first column dtype
print(data.dtypes[0])
print("-----"*15)
# Access the last column dtype
print(data.dtypes[-1])

policyID                int64
statecode              object
county                 object
eq_site_limit         float64
hu_site_limit         float64
fl_site_limit         float64
fr_site_limit         float64
tiv_2011              float64
tiv_2012              float64
eq_site_deductible    float64
hu_site_deductible    float64
fl_site_deductible    float64
fr_site_deductible    float64
point_latitude        float64
point_longitude       float64
line                   object
construction           object
point_granularity     float64
dtype: object
---------------------------------------------------------------------------
<class 'pandas.core.series.Series'>
---------------------------------------------------------------------------
int64
---------------------------------------------------------------------------
float64


In [0]:
# Column count for each datatype
# How many columns are there with int64
# How many columns are there with float64
# How many columns are there with object
data.dtypes.value_counts()
# Note We are using series method value_counts

float64    13
object      4
int64       1
dtype: int64

**observe the data**

In [0]:
# display top 5 rows
data.head()

policyID statecode       county  ...         line  construction  point_granularity
0    119736        FL  CLAY COUNTY  ...  Residential       Masonry                1.0
1    448094        FL  CLAY COUNTY  ...  Residential       Masonry                3.0
2    206893        FL  CLAY COUNTY  ...  Residential          Wood                1.0
3    333743        FL  CLAY COUNTY  ...  Residential          Wood                3.0
4    172534        FL  CLAY COUNTY  ...  Residential          Wood                1.0

[5 rows x 18 columns]

In [0]:
# display top 10 rows
data.head(10)

policyID statecode  ...         construction  point_granularity
0    119736        FL  ...              Masonry                1.0
1    448094        FL  ...              Masonry                3.0
2    206893        FL  ...                 Wood                1.0
3    333743        FL  ...                 Wood                3.0
4    172534        FL  ...                 Wood                1.0
5    785275        FL  ...              Masonry                3.0
6    995932        FL  ...  Reinforced Concrete                1.0
7    223488        FL  ...                 Wood                1.0
8    433512        FL  ...                 Wood                1.0
9    142071        FL  ...              Masonry                1.0

[10 rows x 18 columns]

In [0]:
# display last 5 rows
data.tail()

policyID statecode  ... construction  point_granularity
27883    329719        FL  ...      Masonry                3.0
27884    419086        FL  ...      Masonry                1.0
27885    390281        FL  ...         Wood                1.0
27886    795895        FL  ...         Wood                1.0
27887    891450        FL  ...          NaN                NaN

[5 rows x 18 columns]

In [0]:
# display last 3 rows
data.tail(3)

policyID statecode  ... construction  point_granularity
27885    390281        FL  ...         Wood                1.0
27886    795895        FL  ...         Wood                1.0
27887    891450        FL  ...          NaN                NaN

[3 rows x 18 columns]

In [0]:
# select a random row
data.sample()

policyID statecode  ... construction  point_granularity
10110    584214        FL  ...         Wood                3.0

[1 rows x 18 columns]

In [0]:
# select 5 rows randomly
data.sample(5)

policyID statecode  ... construction  point_granularity
20140    147097        FL  ...         Wood                1.0
19358    693658        FL  ...         Wood                3.0
23697    576131        FL  ...         Wood                1.0
22610    362410        FL  ...         Wood                1.0
13773    685236        FL  ...         Wood                1.0

[5 rows x 18 columns]

- data.head()

- data.tail()

- data.sample()

the output type of all the above methods are dataframe

In [0]:
data.head(1)

policyID statecode       county  ...         line  construction  point_granularity
0    119736        FL  CLAY COUNTY  ...  Residential       Masonry                1.0

[1 rows x 18 columns]

In [0]:
data.tail(1)

policyID statecode county  ...  line  construction  point_granularity
27887    891450        FL    PIN  ...   NaN           NaN                NaN

[1 rows x 18 columns]

- data.head(1)
- data.samle(1)
  - the output type of all the above methods are also dataframe which consists a single row

Info about the dataframe

In [0]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27888 entries, 0 to 27887
Data columns (total 18 columns):
policyID              27888 non-null int64
statecode             27888 non-null object
county                27888 non-null object
eq_site_limit         27887 non-null float64
hu_site_limit         27887 non-null float64
fl_site_limit         27887 non-null float64
fr_site_limit         27887 non-null float64
tiv_2011              27887 non-null float64
tiv_2012              27887 non-null float64
eq_site_deductible    27887 non-null float64
hu_site_deductible    27887 non-null float64
fl_site_deductible    27887 non-null float64
fr_site_deductible    27887 non-null float64
point_latitude        27887 non-null float64
point_longitude       27887 non-null float64
line                  27887 non-null object
construction          27887 non-null object
point_granularity     27887 non-null float64
dtypes: float64(13), int64(1), object(4)
memory usage: 3.8+ MB


In [0]:
type(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27888 entries, 0 to 27887
Data columns (total 18 columns):
policyID              27888 non-null int64
statecode             27888 non-null object
county                27888 non-null object
eq_site_limit         27887 non-null float64
hu_site_limit         27887 non-null float64
fl_site_limit         27887 non-null float64
fr_site_limit         27887 non-null float64
tiv_2011              27887 non-null float64
tiv_2012              27887 non-null float64
eq_site_deductible    27887 non-null float64
hu_site_deductible    27887 non-null float64
fl_site_deductible    27887 non-null float64
fr_site_deductible    27887 non-null float64
point_latitude        27887 non-null float64
point_longitude       27887 non-null float64
line                  27887 non-null object
construction          27887 non-null object
point_granularity     27887 non-null float64
dtypes: float64(13), int64(1), object(4)
memory usage: 3.8+ MB


NoneType

Accesing a single column

In [0]:
# To access the column policyID
data.policyID

0        119736
1        448094
2        206893
3        333743
4        172534
          ...  
27883    329719
27884    419086
27885    390281
27886    795895
27887    891450
Name: policyID, Length: 27888, dtype: int64

In [0]:
# What is the ouput type
print(type(data.policyID))

<class 'pandas.core.series.Series'>


In [0]:
# To access the column policyID
data['policyID']

0        119736
1        448094
2        206893
3        333743
4        172534
          ...  
27883    329719
27884    419086
27885    390281
27886    795895
27887    891450
Name: policyID, Length: 27888, dtype: int64

In [0]:
# What is the ouput type
print(type(data['policyID']))

<class 'pandas.core.series.Series'>


**specific column Datatype**

In [0]:
data.statecode

0        FL
1        FL
2        FL
3        FL
4        FL
         ..
27883    FL
27884    FL
27885    FL
27886    FL
27887    FL
Name: statecode, Length: 27888, dtype: object

In [0]:
# Specific column datatype
data.statecode.dtype

dtype('O')

In [0]:
data.statecode.dtypes

dtype('O')

In [0]:
data.dtype

AttributeError: ignored

- difference b/n dtype and dtypes

- datatafrme object doesn't have dtype
- Numerical columns - int or float data
  - mean,median,mode,sum.quartiles,min,max
- Categorical columns - Object type(string data)
- Bollean data - True or False


In [0]:
data.point_granularity

0        1.0
1        3.0
2        1.0
3        3.0
4        1.0
        ... 
27883    3.0
27884    1.0
27885    1.0
27886    1.0
27887    NaN
Name: point_granularity, Length: 27888, dtype: float64

In [0]:
data.point_granularity.dtype

dtype('float64')

In [0]:
# mean value of a column
data.point_granularity.mean()

1.637106895686162

In [0]:
# mean value of a column
data.point_granularity.sum()/data.point_granularity.shape[0]

1.6370481927710843

In [0]:
# minimum value of a column
print(data.point_granularity.min())
# maximum value of a column
print(data.point_granularity.max())

1.0
4.0


In [0]:
data.point_granularity.value_counts()

1.0    20354
3.0     4832
4.0     2701
Name: point_granularity, dtype: int64

In [0]:
data.point_granularity.mode()
# output type is a series

0    1.0
dtype: float64

In [0]:
data.point_granularity.mode()[0]

1.0

In [0]:
data.point_granularity.median()

1.0

In [0]:
# Range of values in a column
(data.point_granularity.min(),data.point_granularity.max())

(1.0, 4.0)

In [0]:
data.point_granularity.count()

27887

In [0]:
data.county.dtype

dtype('O')

In [0]:
# How many non null values are there
data.county.count()
# Output type is an integer

27888

In [0]:
# For each uniwue value what is the count
data.county.value_counts()
# output type is a series
# In that series the labels are unique values
# the value for that label is how many times it occured in that column
# the final dtype refers to the value type of the series. count is integer

MIAMI DADE COUNTY      4315
BROWARD COUNTY         3193
PALM BEACH COUNTY      2791
DUVAL COUNTY           1894
ORANGE COUNTY          1811
VOLUSIA COUNTY         1367
HILLSBOROUGH COUNTY    1165
OKALOOSA COUNTY        1115
SEMINOLE COUNTY        1100
ALACHUA COUNTY          973
BREVARD COUNTY          872
SANTA ROSA COUNTY       856
ST  JOHNS COUNTY        657
ESCAMBIA COUNTY         494
BAY COUNTY              403
INDIAN RIVER COUNTY     380
PASCO COUNTY            365
CLAY COUNTY             363
WALTON COUNTY           288
PUTNAM COUNTY           268
LEON COUNTY             246
JACKSON COUNTY          208
LAKE COUNTY             206
FLAGLER COUNTY          204
GADSDEN COUNTY          196
SUWANNEE COUNTY         154
SUMTER COUNTY           153
MONROE COUNTY           152
NASSAU COUNTY           135
COLUMBIA COUNTY         125
WASHINGTON COUNTY       116
LEVY COUNTY             114
TAYLOR COUNTY           113
MARTIN COUNTY           109
PINELLAS COUNTY          87
WAKULLA COUNTY      

In [0]:
# How many unique values are there
print(data.county.nunique())
print(data.county.value_counts().shape[0])

55
55


In [0]:
# What are the uinique values
print(data.county.unique())
print(data.county.value_counts().index)

['CLAY COUNTY' 'SUWANNEE COUNTY' 'NASSAU COUNTY' 'COLUMBIA COUNTY'
 'ST  JOHNS COUNTY' 'BAKER COUNTY' 'BRADFORD COUNTY' 'HAMILTON COUNTY'
 'UNION COUNTY' 'MADISON COUNTY' 'LAFAYETTE COUNTY' 'FLAGLER COUNTY'
 'DUVAL COUNTY' 'LAKE COUNTY' 'VOLUSIA COUNTY' 'PUTNAM COUNTY'
 'MARION COUNTY' 'SUMTER COUNTY' 'LEON COUNTY' 'FRANKLIN COUNTY'
 'LIBERTY COUNTY' 'GADSDEN COUNTY' 'WAKULLA COUNTY' 'JEFFERSON COUNTY'
 'TAYLOR COUNTY' 'BAY COUNTY' 'WALTON COUNTY' 'JACKSON COUNTY'
 'CALHOUN COUNTY' 'HOLMES COUNTY' 'WASHINGTON COUNTY' 'GULF COUNTY'
 'ESCAMBIA COUNTY' 'SANTA ROSA COUNTY' 'OKALOOSA COUNTY' 'ALACHUA COUNTY'
 'GILCHRIST COUNTY' 'LEVY COUNTY' 'DIXIE COUNTY' 'SEMINOLE COUNTY'
 'ORANGE COUNTY' 'BREVARD COUNTY' 'INDIAN RIVER COUNTY'
 'MIAMI DADE COUNTY' 'BROWARD COUNTY' 'MONROE COUNTY' 'PALM BEACH COUNTY'
 'MARTIN COUNTY' 'HENDRY COUNTY' 'PASCO COUNTY' 'GLADES COUNTY'
 'HILLSBOROUGH COUNTY' 'HERNANDO COUNTY' 'PINELLAS COUNTY' 'PIN']
Index(['MIAMI DADE COUNTY', 'BROWARD COUNTY', 'PALM BEACH COUN

In [0]:
# To get the description for the numerical columns
data.describe()

policyID  eq_site_limit  ...  point_longitude  point_granularity
count   27888.000000   2.788700e+04  ...     27887.000000       27887.000000
mean   549669.732824   8.270834e+05  ...       -81.817304           1.637107
std    260235.493394   1.990538e+07  ...         2.052113           1.076529
min    100074.000000   0.000000e+00  ...       -87.447290           1.000000
25%    324875.500000   0.000000e+00  ...       -82.308300           1.000000
50%    549945.000000   0.000000e+00  ...       -81.289370           1.000000
75%    775011.500000   0.000000e+00  ...       -80.280120           3.000000
max    999971.000000   2.160000e+09  ...       -80.033257           4.000000

[8 rows x 14 columns]

In [0]:
# To get the list of numerical columns
data.describe().columns

Index(['policyID', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit',
       'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible',
       'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible',
       'point_latitude', 'point_longitude', 'point_granularity'],
      dtype='object')

In [0]:
# To get the description for the categorical columns
data.describe(include=object)

statecode             county         line construction
count      27888              27888        27887        27887
unique         1                 55            2            5
top           FL  MIAMI DADE COUNTY  Residential         Wood
freq       27888               4315        23388        16524

In [0]:
data.describe(include=np.int64)

policyID
count   27888.000000
mean   549669.732824
std    260235.493394
min    100074.000000
25%    324875.500000
50%    549945.000000
75%    775011.500000
max    999971.000000

In [0]:
data.describe(include=np.float64)

eq_site_limit  hu_site_limit  ...  point_longitude  point_granularity
count   2.788700e+04   2.788700e+04  ...     27887.000000       27887.000000
mean    8.270834e+05   2.270114e+06  ...       -81.817304           1.637107
std     1.990538e+07   2.220591e+07  ...         2.052113           1.076529
min     0.000000e+00   0.000000e+00  ...       -87.447290           1.000000
25%     0.000000e+00   3.816614e+04  ...       -82.308300           1.000000
50%     0.000000e+00   1.805753e+05  ...       -81.289370           1.000000
75%     0.000000e+00   9.215907e+05  ...       -80.280120           3.000000
max     2.160000e+09   2.160000e+09  ...       -80.033257           4.000000

[8 rows x 13 columns]

In [0]:
data.describe(include='all')

policyID statecode  ... construction  point_granularity
count    27888.000000     27888  ...        27887       27887.000000
unique            NaN         1  ...            5                NaN
top               NaN        FL  ...         Wood                NaN
freq              NaN     27888  ...        16524                NaN
mean    549669.732824       NaN  ...          NaN           1.637107
std     260235.493394       NaN  ...          NaN           1.076529
min     100074.000000       NaN  ...          NaN           1.000000
25%     324875.500000       NaN  ...          NaN           1.000000
50%     549945.000000       NaN  ...          NaN           1.000000
75%     775011.500000       NaN  ...          NaN           3.000000
max     999971.000000       NaN  ...          NaN           4.000000

[11 rows x 18 columns]

In [0]:
data.columns

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')

- Slicing, subsetting, Boolean Indexing

- postion based (iloc)
   - rows --> 0 to 36633
   - columns --> 0 to 17
- label based indexing(loc)
- rows --> 0 to 36633
- columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible', 'point_latitude', 'point_longitude', 'line', 'construction', 'point_granularity']

In [0]:
data.iloc[1,:] # row with index 1 and all the columns

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [0]:
data.loc[1,:]

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [0]:
data.iloc[1,1] # row with index 1and column with index 1

'FL'

In [0]:
data.loc[1,"statecode"]

'FL'

In [0]:
data.iloc[1,1:3] # row with index 1 and columns with index 1,2
# iloc stop index is exclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [0]:
data.loc[1,"statecode":"county"] 
# loc - the stop label also inclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [0]:
data.iloc[1:4,1:3] 
# rows with indices 1,2,3
# columns with indices 1,2

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [0]:
data.loc[1:3,"statecode":"county"]

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [0]:
data.policyID.nunique()

27888

In [0]:
data.set_index("policyID")

statecode           county  ...  construction  point_granularity
policyID                             ...                                 
119736          FL      CLAY COUNTY  ...       Masonry                1.0
448094          FL      CLAY COUNTY  ...       Masonry                3.0
206893          FL      CLAY COUNTY  ...          Wood                1.0
333743          FL      CLAY COUNTY  ...          Wood                3.0
172534          FL      CLAY COUNTY  ...          Wood                1.0
...            ...              ...  ...           ...                ...
329719          FL  PINELLAS COUNTY  ...       Masonry                3.0
419086          FL  PINELLAS COUNTY  ...       Masonry                1.0
390281          FL  PINELLAS COUNTY  ...          Wood                1.0
795895          FL  PINELLAS COUNTY  ...          Wood                1.0
891450          FL              PIN  ...           NaN                NaN

[27888 rows x 17 columns]

In [0]:
data

policyID statecode  ... construction  point_granularity
0        119736        FL  ...      Masonry                1.0
1        448094        FL  ...      Masonry                3.0
2        206893        FL  ...         Wood                1.0
3        333743        FL  ...         Wood                3.0
4        172534        FL  ...         Wood                1.0
...         ...       ...  ...          ...                ...
27883    329719        FL  ...      Masonry                3.0
27884    419086        FL  ...      Masonry                1.0
27885    390281        FL  ...         Wood                1.0
27886    795895        FL  ...         Wood                1.0
27887    891450        FL  ...          NaN                NaN

[27888 rows x 18 columns]

In [0]:
df = data.set_index("policyID")
df.head()

statecode       county  ...  construction  point_granularity
policyID                         ...                                 
119736          FL  CLAY COUNTY  ...       Masonry                1.0
448094          FL  CLAY COUNTY  ...       Masonry                3.0
206893          FL  CLAY COUNTY  ...          Wood                1.0
333743          FL  CLAY COUNTY  ...          Wood                3.0
172534          FL  CLAY COUNTY  ...          Wood                1.0

[5 rows x 17 columns]

In [0]:
df.index

Int64Index([119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
            433512, 142071,
            ...
            902096, 291536, 213120, 363687, 366207, 329719, 419086, 390281,
            795895, 891450],
           dtype='int64', name='policyID', length=27888)

- Slicing, subsetting, Boolean Indexing with df

- postion based (iloc)
  - rows --> 0 to 36633
  - columns --> 0 to 17
- label based indexing(loc)
- rows --> [119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
  433512, 142071,...]
- columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible', 'point_latitude', 'point_longitude', 'line', 'construction', 'point_granularity']

In [0]:
# row with index 1 and all the columns

In [0]:
df.iloc[1,:]

statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 448094, dtype: object

In [0]:
df.loc[448094,:]

statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 448094, dtype: object

In [0]:
# row with index 1and column with index 1

In [0]:
df.iloc[1,1]

'CLAY COUNTY'

In [0]:
df.loc[448094,"county"]

'CLAY COUNTY'

In [0]:
# row with index 1 and columns with index 1,2

In [0]:
df.iloc[1,1:3]

county           CLAY COUNTY
eq_site_limit    1.32238e+06
Name: 448094, dtype: object

In [0]:
df.loc[ 448094,"county":"eq_site_limit"]

county           CLAY COUNTY
eq_site_limit    1.32238e+06
Name: 448094, dtype: object

In [0]:
# rows with indices 1,2,3
# columns with indices 1,2

In [0]:
df.iloc[1:4,:]

statecode       county  ...  construction  point_granularity
policyID                         ...                                 
448094          FL  CLAY COUNTY  ...       Masonry                3.0
206893          FL  CLAY COUNTY  ...          Wood                1.0
333743          FL  CLAY COUNTY  ...          Wood                3.0

[3 rows x 17 columns]

In [0]:
df.loc[448094:333743,:]

statecode       county  ...  construction  point_granularity
policyID                         ...                                 
448094          FL  CLAY COUNTY  ...       Masonry                3.0
206893          FL  CLAY COUNTY  ...          Wood                1.0
333743          FL  CLAY COUNTY  ...          Wood                3.0

[3 rows x 17 columns]

**Boolean Indexing**

In [0]:
df.dtypes

statecode              object
county                 object
eq_site_limit         float64
hu_site_limit         float64
fl_site_limit         float64
fr_site_limit         float64
tiv_2011              float64
tiv_2012              float64
eq_site_deductible    float64
hu_site_deductible    float64
fl_site_deductible    float64
fr_site_deductible    float64
point_latitude        float64
point_longitude       float64
line                   object
construction           object
point_granularity     float64
dtype: object

In [0]:
df.dtypes == object

statecode              True
county                 True
eq_site_limit         False
hu_site_limit         False
fl_site_limit         False
fr_site_limit         False
tiv_2011              False
tiv_2012              False
eq_site_deductible    False
hu_site_deductible    False
fl_site_deductible    False
fr_site_deductible    False
point_latitude        False
point_longitude       False
line                   True
construction           True
point_granularity     False
dtype: bool

In [0]:
df.loc[:,df.dtypes == object]

statecode           county         line construction
policyID                                                     
119736          FL      CLAY COUNTY  Residential      Masonry
448094          FL      CLAY COUNTY  Residential      Masonry
206893          FL      CLAY COUNTY  Residential         Wood
333743          FL      CLAY COUNTY  Residential         Wood
172534          FL      CLAY COUNTY  Residential         Wood
...            ...              ...          ...          ...
329719          FL  PINELLAS COUNTY  Residential      Masonry
419086          FL  PINELLAS COUNTY  Residential      Masonry
390281          FL  PINELLAS COUNTY  Residential         Wood
795895          FL  PINELLAS COUNTY  Residential         Wood
891450          FL              PIN          NaN          NaN

[27888 rows x 4 columns]

In [0]:
df.dtypes ==np.int64

statecode             False
county                False
eq_site_limit         False
hu_site_limit         False
fl_site_limit         False
fr_site_limit         False
tiv_2011              False
tiv_2012              False
eq_site_deductible    False
hu_site_deductible    False
fl_site_deductible    False
fr_site_deductible    False
point_latitude        False
point_longitude       False
line                  False
construction          False
point_granularity     False
dtype: bool

In [0]:
df.loc[:,df.dtypes == np.int64]

Empty DataFrame
Columns: []
Index: [119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488, 433512, 142071, 253816, 894922, 422834, 582721, 842700, 874333, 580146, 456149, 767862, 353022, 367814, 671392, 772887, 983122, 934215, 385951, 716332, 751262, 633663, 105851, 710400, 703001, 352792, 717603, 937659, 294022, 410500, 524433, 779298, 491831, 814637, 737515, 222653, 788543, 691681, 368807, 174002, 198760, 831395, 305698, 515722, 415582, 783533, 640802, 403866, 828788, 751490, 972562, 367541, 481360, 920232, 727659, 471817, 983043, 578286, 347276, 477786, 265907, 107560, 908183, 356024, 575299, 790478, 801776, 291781, 537286, 289305, 697409, 633944, 963511, 729649, 862584, 371891, 785595, 567696, 219025, 447216, 305141, 625575, 262109, 228457, 622799, 233271, 932950, 641952, 796260, 301822, 559927, 224400, 123936, ...]

[27888 rows x 0 columns]

In [0]:
df.county

policyID
119736        CLAY COUNTY
448094        CLAY COUNTY
206893        CLAY COUNTY
333743        CLAY COUNTY
172534        CLAY COUNTY
               ...       
329719    PINELLAS COUNTY
419086    PINELLAS COUNTY
390281    PINELLAS COUNTY
795895    PINELLAS COUNTY
891450                PIN
Name: county, Length: 27888, dtype: object

In [0]:
df.county.nunique()

55

In [0]:
df.county.unique()

array(['CLAY COUNTY', 'SUWANNEE COUNTY', 'NASSAU COUNTY',
       'COLUMBIA COUNTY', 'ST  JOHNS COUNTY', 'BAKER COUNTY',
       'BRADFORD COUNTY', 'HAMILTON COUNTY', 'UNION COUNTY',
       'MADISON COUNTY', 'LAFAYETTE COUNTY', 'FLAGLER COUNTY',
       'DUVAL COUNTY', 'LAKE COUNTY', 'VOLUSIA COUNTY', 'PUTNAM COUNTY',
       'MARION COUNTY', 'SUMTER COUNTY', 'LEON COUNTY', 'FRANKLIN COUNTY',
       'LIBERTY COUNTY', 'GADSDEN COUNTY', 'WAKULLA COUNTY',
       'JEFFERSON COUNTY', 'TAYLOR COUNTY', 'BAY COUNTY', 'WALTON COUNTY',
       'JACKSON COUNTY', 'CALHOUN COUNTY', 'HOLMES COUNTY',
       'WASHINGTON COUNTY', 'GULF COUNTY', 'ESCAMBIA COUNTY',
       'SANTA ROSA COUNTY', 'OKALOOSA COUNTY', 'ALACHUA COUNTY',
       'GILCHRIST COUNTY', 'LEVY COUNTY', 'DIXIE COUNTY',
       'SEMINOLE COUNTY', 'ORANGE COUNTY', 'BREVARD COUNTY',
       'INDIAN RIVER COUNTY', 'MIAMI DADE COUNTY', 'BROWARD COUNTY',
       'MONROE COUNTY', 'PALM BEACH COUNTY', 'MARTIN COUNTY',
       'HENDRY COUNTY', 'PASCO COUN

In [0]:
df.county== 'CLAY COUNTY'

policyID
119736     True
448094     True
206893     True
333743     True
172534     True
          ...  
329719    False
419086    False
390281    False
795895    False
891450    False
Name: county, Length: 27888, dtype: bool

In [0]:
df.loc[df.county== 'CLAY COUNTY',:]

statecode       county  ...  construction  point_granularity
policyID                         ...                                 
119736          FL  CLAY COUNTY  ...       Masonry                1.0
448094          FL  CLAY COUNTY  ...       Masonry                3.0
206893          FL  CLAY COUNTY  ...          Wood                1.0
333743          FL  CLAY COUNTY  ...          Wood                3.0
172534          FL  CLAY COUNTY  ...          Wood                1.0
...            ...          ...  ...           ...                ...
767514          FL  CLAY COUNTY  ...          Wood                1.0
632481          FL  CLAY COUNTY  ...          Wood                1.0
703886          FL  CLAY COUNTY  ...          Wood                1.0
992295          FL  CLAY COUNTY  ...          Wood                1.0
662150          FL  CLAY COUNTY  ...          Wood                1.0

[363 rows x 17 columns]

**Access Multiple columns**

In [0]:
df[['county','statecode','point_granularity']]

county statecode  point_granularity
policyID                                              
119736        CLAY COUNTY        FL                1.0
448094        CLAY COUNTY        FL                3.0
206893        CLAY COUNTY        FL                1.0
333743        CLAY COUNTY        FL                3.0
172534        CLAY COUNTY        FL                1.0
...                   ...       ...                ...
329719    PINELLAS COUNTY        FL                3.0
419086    PINELLAS COUNTY        FL                1.0
390281    PINELLAS COUNTY        FL                1.0
795895    PINELLAS COUNTY        FL                1.0
891450                PIN        FL                NaN

[27888 rows x 3 columns]

In [0]:
df[['county']]

county
policyID                 
119736        CLAY COUNTY
448094        CLAY COUNTY
206893        CLAY COUNTY
333743        CLAY COUNTY
172534        CLAY COUNTY
...                   ...
329719    PINELLAS COUNTY
419086    PINELLAS COUNTY
390281    PINELLAS COUNTY
795895    PINELLAS COUNTY
891450                PIN

[27888 rows x 1 columns]

In [0]:
df['county'] - It is a series
df[['county']] - DataFrame

Access row with index 133 as a series

In [0]:
df.iloc[133]

statecode                           FL
county                ST  JOHNS COUNTY
eq_site_limit                        0
hu_site_limit                   159042
fl_site_limit                        0
fr_site_limit                        0
tiv_2011                        159042
tiv_2012                        249985
eq_site_deductible                   0
hu_site_deductible                   0
fl_site_deductible                   0
fr_site_deductible                   0
point_latitude                 29.7766
point_longitude               -81.4717
line                       Residential
construction                      Wood
point_granularity                    1
Name: 482457, dtype: object

In [0]:
df.iloc[[133]]

statecode            county  ...  construction  point_granularity
policyID                              ...                                 
482457          FL  ST  JOHNS COUNTY  ...          Wood                1.0

[1 rows x 17 columns]

In [0]:
df.iloc[133:134]

statecode            county  ...  construction  point_granularity
policyID                              ...                                 
482457          FL  ST  JOHNS COUNTY  ...          Wood                1.0

[1 rows x 17 columns]

In [0]:
df.county.isnull()

policyID
119736    False
448094    False
206893    False
333743    False
172534    False
          ...  
329719    False
419086    False
390281    False
795895    False
891450    False
Name: county, Length: 27888, dtype: bool

In [0]:
df.county.isna()

policyID
119736    False
448094    False
206893    False
333743    False
172534    False
          ...  
329719    False
419086    False
390281    False
795895    False
891450    False
Name: county, Length: 27888, dtype: bool

In [0]:
sample = pd.read_csv("/sample.csv")
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
sample.salary.isna()

0    False
1    False
2    False
3    False
Name: salary, dtype: bool

In [0]:
sample.salary.isnull()

0    False
1    False
2    False
3    False
Name: salary, dtype: bool

In [0]:
sample.isnull()

name  salary   year
0  False   False  False
1  False   False  False
2  False   False  False
3  False   False  False

In [0]:
sample.isna()

name  salary   year
0  False   False  False
1  False   False  False
2  False   False  False
3  False   False  False

In [0]:
True+False+True # Number of true

2

In [0]:
sample.isna().sum() # default axis=0

name      0
salary    0
year      0
dtype: int64

In [0]:
sample.isna().sum(axis=1)

0    0
1    0
2    0
3    0
dtype: int64

In [0]:
sample[sample.salary.isna()]

Empty DataFrame
Columns: [name, salary, year]
Index: []

In [0]:
sample.loc[sample.salary.isna()]

Empty DataFrame
Columns: [name, salary, year]
Index: []

To remove columns

In [0]:
# To remove a column with label salary
sample.drop("salary",axis=1)
# inplace =TRue will overwrite the original column
# sample = sample.drop("salary",axis=1)

name  year
0      Raj Acharya  2015
1      SaiTeja Ram  2012
2    Hemendra Arya  2018
3  Revanth Sundarm  2019

In [0]:
# sample.drop("salary")
# sample.drop("hike",axis=1)

In [0]:
# To remove columns with labels salary and year
sample.drop(["salary",'year'],axis=1)

name
0      Raj Acharya
1      SaiTeja Ram
2    Hemendra Arya
3  Revanth Sundarm

In [0]:
# To remove columns with indices 1 and 2
sample.drop(index = [1,2],axis=1)

name  salary  year
0      Raj Acharya   45000  2015
3  Revanth Sundarm   15000  2019

In [0]:
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
help(sample.drop)

Help on method drop in module pandas.core.frame:

drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise') method of pandas.core.frame.DataFrame instance
    Drop specified labels from rows or columns.
    
    Remove rows or columns by specifying label names and corresponding
    axis, or by specifying directly index or column names. When using a
    multi-index, labels on different levels can be removed by specifying
    the level.
    
    Parameters
    ----------
    labels : single label or list-like
        Index or column labels to drop.
    axis : {0 or 'index', 1 or 'columns'}, default 0
        Whether to drop labels from the index (0 or 'index') or
        columns (1 or 'columns').
    index : single label or list-like
        Alternative to specifying axis (``labels, axis=0``
        is equivalent to ``index=labels``).
    
        .. versionadded:: 0.21.0
    columns : single label or list-like
        Alternative to specifying axis (

**To remove rows**

In [0]:
# To remove rows with lables 0 and 1
sample.drop([0,1],axis=0)

name  salary  year
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
# To remove rows with indices 0 and 1
sample.drop(index = [0,1],axis=0)

name  salary  year
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
# To remove a single row with index 3
sample.drop(3,axis=0)

name  salary  year
0    Raj Acharya   45000  2015
1    SaiTeja Ram   78000  2012
2  Hemendra Arya   32000  2018

In [0]:
#sample.drop(index=5,axis=0)

In [0]:
sample.dropna() # Removes the rows where we have missing values

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
sample.dropna(axis=1) # Removes the columns where we have missing values

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
sample.salary.dropna()

0    45000
1    78000
2    32000
3    15000
Name: salary, dtype: int64

**To fill the missing value with 0**

In [0]:
sample.fillna(0)

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
sample.salary.fillna(0)
# sample.salary = sample.salary.fillna(0)

0    45000
1    78000
2    32000
3    15000
Name: salary, dtype: int64

**Fill the missing value with mean**

In [0]:
m =  sample.salary.mean()
print(m)
sample.salary.fillna(m)

42500.0


0    45000
1    78000
2    32000
3    15000
Name: salary, dtype: int64

In [0]:
# Total Missing value count in the data
sample.isnull().sum().sum()

0

In [0]:
# To get misisng value count in a column
sample.salary.isnull().sum()

0

In [0]:
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
sample_out = sample.copy()
sample_out['salary'] = sample_out.salary.fillna(sample_out.salary.mean())
sample_out

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
sample_out_rows = sample.dropna()
sample_out_rows

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
sample_out_cols = sample.dropna(axis=1)
sample_out_cols

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
dummy = sample.copy()
dummy

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
dummy.salary = dummy.salary.mean()
dummy

name   salary  year
0      Raj Acharya  42500.0  2015
1      SaiTeja Ram  42500.0  2012
2    Hemendra Arya  42500.0  2018
3  Revanth Sundarm  42500.0  2019

In [0]:
dummy = sample.copy()
dummy

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
dummy.salary = dummy.salary.mean
dummy

name                                             salary  year
0      Raj Acharya  <bound method Series.mean of 0    45000\n1    ...  2015
1      SaiTeja Ram  <bound method Series.mean of 0    45000\n1    ...  2012
2    Hemendra Arya  <bound method Series.mean of 0    45000\n1    ...  2018
3  Revanth Sundarm  <bound method Series.mean of 0    45000\n1    ...  2019

- pd.get_dummies

- It will create dummy columns with unique  
   -  values as column names
   -  One hot encoded data

In [0]:
s = pd.Series(["avg","good","bad","avg","good"])
pd.get_dummies(s)

avg  bad  good
0    1    0     0
1    0    0     1
2    0    1     0
3    1    0     0
4    0    0     1

In [0]:
s.nunique()

3

In [0]:
s.unique()

array(['avg', 'good', 'bad'], dtype=object)

In [0]:
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [0]:
pd.get_dummies(sample)

salary  year  ...  name_Revanth Sundarm  name_SaiTeja Ram
0   45000  2015  ...                     0                 0
1   78000  2012  ...                     0                 1
2   32000  2018  ...                     0                 0
3   15000  2019  ...                     1                 0

[4 rows x 6 columns]

In [0]:
data.dtypes==object

policyID              False
statecode              True
county                 True
eq_site_limit         False
hu_site_limit         False
fl_site_limit         False
fr_site_limit         False
tiv_2011              False
tiv_2012              False
eq_site_deductible    False
hu_site_deductible    False
fl_site_deductible    False
fr_site_deductible    False
point_latitude        False
point_longitude       False
line                   True
construction           True
point_granularity     False
dtype: bool

In [0]:
data.loc[:,data.dtypes==object].nunique()

statecode        1
county          36
line             2
construction     5
dtype: int64

In [0]:
data.loc[:,data.dtypes==object].nunique().sum()

44

In [0]:
catageroical_columns_count = (data.dtypes==object).sum()
catageroical_columns_count

4

In [0]:
# no of columns in the output
75 - 4 + 18

89

In [0]:
pd.get_dummies(data)

policyID  eq_site_limit  ...  construction_Steel Frame  construction_Wood
0       119736       498960.0  ...                         0                  0
1       448094      1322376.3  ...                         0                  0
2       206893       190724.4  ...                         0                  1
3       333743            0.0  ...                         0                  1
4       172534            0.0  ...                         0                  1
...        ...            ...  ...                       ...                ...
9457    441177            0.0  ...                         0                  1
9458    264124            0.0  ...                         0                  1
9459    494493            0.0  ...                         0                  1
9460    972124            0.0  ...                         0                  1
9461    566824            0.0  ...                         0                  0

[9462 rows x 58 columns]

In [0]:
pd.get_dummies(data).shape[1]

58